packages/runtime/src/graph.ts

```diff
@@ -20,12 +20,76 @@ export class Graph {
 		return new Graph({ id });
 	}
 
-	static async new(..._args: tg.Args<Graph.Arg>): Promise<Graph> {
-		return todo();
+	static async new(...args: tg.Args<Graph.Arg>): Promise<Graph> {
+		// Combine incoming arguments.
+		let arg = await Graph.arg(...args);
+
+		// Make all optional argument fields explicit.
+		let nodes: Array<Graph.Node> = await Promise.all(
+			(arg.nodes ?? []).map(async (node) => {
+				if (node.kind === "directory") {
+					return {
+						kind: "directory",
+						entries: node.entries ?? {},
+					};
+				} else if (node.kind === "file") {
+					return {
+						kind: "file",
+						contents: await tg.blob(node.contents),
+						dependencies: node.dependencies ?? undefined,
+						executable: node.executable ?? false,
+					};
+				} else if (node.kind === "symlink") {
+					return {
+						kind: "symlink",
+						artifact: node.artifact ?? undefined,
+						path: node.path ? tg.path(node.path) : undefined,
+					};
+				} else {
+					return tg.unreachable(node);
+				}
+			})
+		);
+
+		// Construct the object and return a new instance.
+		return new Graph({ object: { nodes } });
 	}
 
-	static async arg(..._args: tg.Args<Graph.Arg>): Promise<Graph.ArgObject> {
-		return todo();
+	static async arg(...args: tg.Args<Graph.Arg>): Promise<Graph.ArgObject> {
+		// Resolve and flatten the incoming arguments.
+		let resolved = await Promise.all(args.map(tg.resolve));
+		let flattened = flatten(resolved);
+
+		// Add the given offset to all indices in the object.
+		let addOffset = (obj: any, offset: number): any => {
+			if (typeof obj === "number") return obj + offset;
+			if (Array.isArray(obj)) return obj.map((item) => addOffset(item, offset));
+			if (obj && typeof obj === "object") {
+				return Object.fromEntries(
+					Object.entries(obj).map(([key, value]) => [
+						key,
+						addOffset(value, offset),
+					])
+				);
+			}
+			return obj;
+		};
```

Comment:
Instead of treating the object as "any" and adding the offset to all numbers you discover, can you use the types to add it to only the dependencies, entries, etc.?